contracts/src/v0.8/dev/ocr2dr/AuthorizedOriginReceiver.sol

```diff
@@ -0,0 +1,114 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.6;
+
+import "@openzeppelin/contracts/utils/structs/EnumerableSet.sol";
+import "../interfaces/AuthorizedReceiverInterface.sol";
+
+/**
+ * @notice Modified AuthorizedReciever abstract for use on the OCR2DROracle contract to limit usage
+ * @notice Uses tx.origin instead of msg.sender because the client contract sends messages to the Oracle contract
+ * @dev NOTE: Use the following steps to use for deployments. Do not leave these changes in the repository code.
+ * @dev To use:
+ * 1. Make the Oracle contract ownable, to control who can set the authorized senders
+ * ```
+ * import "../../ConfirmedOwner.sol";
+ * ...
+ * contract OCR2DROracle is OCR2DROracleInterface, OCR2Base, ConfirmedOwner {
+ * ...
+ * constructor() OCR2Base(true) ConfirmedOwner(msg.sender) {}
+ * ```
+ *
+ * 2. Extend OCR2DROracle.sol with this contract
+ * ```
+ * import "./AuthorizedOriginReceiver.sol";
+ *
+ * contract OCR2DROracle is OCR2DROracleInterface, OCR2Base, ConfirmedOwner, AuthorizedOriginReceiver {
+ * ```
+ *
+ * 3. Override the virtual function _canSetAuthorizedSenders
+ * ```
+ *   function _canSetAuthorizedSenders() internal view override onlyOwner returns (bool) {
+ *   return true;
+ * }
+ * ```
+ */
+abstract contract AuthorizedOriginReceiver is AuthorizedReceiverInterface {
+  using EnumerableSet for EnumerableSet.AddressSet;
+
+  event AuthorizedSendersChanged(address[] senders, address changedBy);
+
+  error EmptySendersList();
+  error UnauthorizedSender();
+  error NotAllowedToSetSenders();
+
+  EnumerableSet.AddressSet private s_authorizedSenders;
+  address[] private s_authorizedSendersList;
+
+  /**
+   * @notice Sets the fulfillment permission for a given node. Use `true` to allow, `false` to disallow.
+   * @param senders The addresses of the authorized Chainlink node
+   */
+  function setAuthorizedSenders(address[] calldata senders) external override validateAuthorizedSenderSetter {
+    if (senders.length == 0) {
+      revert EmptySendersList();
+    }
+    for (uint256 i = 0; i < s_authorizedSendersList.length; i++) {
+      s_authorizedSenders.remove(s_authorizedSendersList[i]);
+    }
```

Comment:
Do we want the `add` and `remove` functionality in a single function?  My concern is that if we have to whitelist may addresses (500+), it will require too much gas for a single tx.

Reply:
IMO, we should have 2 functions. `addAuthorizedSenders(address[] senders)` and `removeAuthorizedSenders(address[] senders)`

Reply:
On second thought, even `removeAuthorizedSenders` might fail if the whitelist gets too large since we would have to loop thru every address to find which senders to remove.

We might have to implement ALL of this logic exclusively w/ a mapping, eliminating the `s_authorizedSendersList` array entirely.  This would eliminate the ability to read a list of approved addresses, but at least would never run into gas problems.

Reply:
tbh, we shouldn't really need an array of approved addresses anyway as we can just keep this data up-to-date in a spreadsheet.
We should also add the ability to delete _every_ mapping in the array and re-add an array of authorized senders.

Reply:
I'll modify to add and remove because of what Bolek mentioned.
But I think that gas concerns are an edge case.
For closed beta I don't think we will have even close to that many users.
We can set expectations with Product if that is a concern.

Reply:
I am not sure how many beta users Bryan is expecting, but I know he was scoping out costs for a minimum of 500 users.

I would rather we support this now rather than have to modify contracts later if we run into problems.

Reply:
Let's have addAuthorizedSenders(address[] senders) and removeAuthorizedSenders(address[] senders)

Reply:
I still think this is early optimizing for what is an extreme edge case. Just a quick google search shows [this stackoverflow](https://stackoverflow.com/questions/48113615/solidity-for-loop-over-a-huge-amount-of-data) talking about 500k iterations.

The indices approach sounds fine if you are still very concerned about iterating `s_authorizedSendersList`.

Reply:
This is probably overkill, but IMO this would be the ideal implementation for gas efficiency. (Although `getIndicesToRemove` is just for convenience and could be removed too.)

`function getIndicesToRemove(address[] calldata sendersToRemove) returns (uint256[]) external view {
    uint256[] memory indices = new uint256[](sendersToRemove.length);
    for (uint256 i = 0; i < sendersToRemove.length; i++) {
      for (uint256 j = 0; j < s_authorizedSendersList.length; j++) {
        if (sendersToRemove[i] == s_authorizedSendersList[j]) {
          indices[i] = j;
          continue;
        }
      }
    }
    return indices
  }

  /**
   * @notice Remove the permission to request for the given wallet(s).
   * @param senders The addresses of the wallet addresses to revoke access
   */
  function removeAuthorizedSenders(address[] calldata senders, uint256[] calldata indices) external validateAuthorizedSenderSetter {
    if (senders.length == 0) {
      revert EmptySendersList();
    }
    require(senders.length == indices.length, "length mismatch");
    for (uint256 i = 0; i < senders.length; i++) {
      // Check that the indices argument is correct
      if (senders[i] !== s_authorizedSendersList[indices[i]]) {
        revert IncorrectIndices();
      }
      s_authorizedSenders.remove(senders[i]);
      // Remove from s_authorizedSendersList
      address last = s_authorizedSendersList[s_authorizedSendersList.length - 1];
      s_authorizedSendersList[indices[i]] = last;
      s_authorizedSendersList.pop();
    }
    emit AuthorizedSendersChanged(senders, msg.sender);
  }`